File: ruby/409_longest_palindrome.rb
# 409. Longest Palindrome
# Easy
#
# Given a string s which consists of lowercase or uppercase letters, return
# the length of the longest palindrome that can be built with those letters.
#
# Letters are case sensitive, for example, "Aa" is not considered a
# palindrome here.

def longest_palindrome(s)
  res = 0
  rem = 0

  counts(s).each do |c|
    if c.even?
      res += c
    else
      res += (c - 1)
      rem = 1
    end
  end

  res + rem
end

def counts(s)
  f = Hash.new(0)
  s.each_char do |c|
    f[c] += 1
  end
  f.values
end

File: python/409_longest_palindrome.py
"""
409. Longest Palindrome
Easy

Given a string s which consists of lowercase or uppercase letters, return
the length of the longest palindrome that can be built with those letters.

Letters are case sensitive, for example, "Aa" is not considered a
palindrome here.
"""


class Solution:
    def longestPalindrome(self, s: str) -> int:
        d = {}
        for c in s:
            d[c] = d.get(c, 0) + 1

        res = 0
        rem = 0
        for v in d.values():
            if v % 2 == 0:
                res += v
            else:
                res += v - 1
                rem = 1

        return res + rem
